zhaquirks/danfoss/thermostat.py

```diff
@@ -26,113 +55,442 @@
-    """Danfoss custom cluster."""
-
-    attributes = Diagnostic.attributes.copy()
-    attributes.update(
-        {
-            0x4000: ("sw_error_code", t.bitmap16, True),
-            0x4001: ("wake_time_avg", t.uint32_t, True),
-            0x4002: ("wake_time_max_duration", t.uint32_t, True),
-            0x4003: ("wake_time_min_duration", t.uint32_t, True),
-            0x4004: ("sleep_postponed_count_avg", t.uint32_t, True),
-            0x4005: ("sleep_postponed_count_max", t.uint32_t, True),
-            0x4006: ("sleep_postponed_count_min", t.uint32_t, True),
-            0x4010: ("motor_step_counter", t.uint32_t, True),
-        }
-    )
+
+class DanfossDiagnosticCluster(CustomizedStandardCluster, Diagnostic):
+    """Danfoss cluster for standard and proprietary danfoss attributes."""
+
+    class AttributeDefs(Diagnostic.AttributeDefs):
+        sw_error_code = ZCLAttributeDef(
+            id=0x4000,
+            type=DanfossSoftwareErrorCodeBitmap,
+            access="rpw",
+            is_manufacturer_specific=True,
+        )
+        wake_time_avg = ZCLAttributeDef(
+            id=0x4001, type=types.uint32_t, access="r", is_manufacturer_specific=True
+        )
+        wake_time_max_duration = ZCLAttributeDef(
+            id=0x4002, type=types.uint32_t, access="r", is_manufacturer_specific=True
+        )
+        wake_time_min_duration = ZCLAttributeDef(
+            id=0x4003, type=types.uint32_t, access="r", is_manufacturer_specific=True
+        )
+        sleep_postponed_count_avg = ZCLAttributeDef(
+            id=0x4004, type=types.uint32_t, access="r", is_manufacturer_specific=True
+        )
+        sleep_postponed_count_max = ZCLAttributeDef(
+            id=0x4005, type=types.uint32_t, access="r", is_manufacturer_specific=True
+        )
+        sleep_postponed_count_min = ZCLAttributeDef(
+            id=0x4006, type=types.uint32_t, access="r", is_manufacturer_specific=True
+        )
+        motor_step_counter = ZCLAttributeDef(
+            id=0x4010, type=types.uint32_t, access="rp", is_manufacturer_specific=True
+        )
+
+        data_logger = ZCLAttributeDef(
+            id=0x4020,
+            type=types.LimitedLVBytes(50),
+            access="rpw",
+            is_manufacturer_specific=True,
+        )
+        control_diagnostics = ZCLAttributeDef(
+            id=0x4021,
+            type=types.LimitedLVBytes(30),
+            access="rp",
+            is_manufacturer_specific=True,
+        )
+        control_diagnostics_frequency = ZCLAttributeDef(
+            id=0x4022, type=types.uint16_t, access="rw", is_manufacturer_specific=True
+        )  # non-configurable reporting
+
+
+class DanfossTimeCluster(CustomizedStandardCluster, Time):
+    """Danfoss cluster for fixing the time."""
+
+    async def write_time(self):
+        epoch = datetime(2000, 1, 1, 0, 0, 0, 0)
+        current_time = (datetime.utcnow() - epoch).total_seconds()
```

Comment:
`utcnow()` is deprecated and replaced by `datetime.datetime.now(datetime.UTC)`
(There might then be issues with the subtraction because of one datetime having a timezone and the other not.)